llvm/utils/TableGen/DAGISelMatcherOpt.cpp

```diff
@@ -492,6 +480,20 @@ static void FactorNodes(std::unique_ptr<Matcher> &InputMatcherPtr) {
     Scope->resetChild(i, OptionsToMatch[i]);
 }
 
+/// Search a ScopeMatcher to factor with FactorScope.
+static void FactorNodes(std::unique_ptr<Matcher> &InputMatcherPtr) {
```

Comment:
I don't see why this is referencing unique_ptr at all. This could be done in terms of Matcher* or Matcher&

Reply:
The code change the contents of the unique_ptr by calling unique_ptr::reset in several places. It replaces the ScopeMatcher with a SwitchOpcodeMatcher/SwitchType matcher by changing what the owner of ScopeMatcher is pointing to. This requires being able to modify the unique_ptr inside the owner.